File: Problems/46_Permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        
        def helper(curr: List[int]):
            if len(curr) == len(nums):
                res.append(curr)
                return
            for num in nums:
                if num in curr:
                    continue
                helper(curr + [num])
        helper([])
        return res

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def helper(currNums, usedIndex):
            if  len(currNums) == len(nums):
                res.append(currNums)
                return
            for i, num in enumerate(nums):
                if i in usedIndex:
                    continue
                helper(currNums + [num], usedIndex + [i])
        
        helper([], [])
        return res
                

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def helper(curr_list, history_index_list):
            if len(curr_list)==len(nums):
                res.append(curr_list)
                return 
            for i in range(len(nums)):
                if i in history_index_list:
                    continue
                helper(curr_list+[nums[i]], history_index_list+[i])
        helper([], [])
        return res

if __name__ == '__main__':
    sol = Solution()
    nums = [1,2,3]
    print(sol.permute(nums))

File: Problems/46/46_Permutations.cpp
class Solution {
public:
    vector<vector<int>> res;
    int MAX_LEN;
    vector<int> nums_;
    vector<vector<int>> permute(vector<int>& nums) {
        MAX_LEN = nums.size();
        nums_ = nums;
        vector<int> tmp1;
        vector<int> tmp2;
        helper(tmp1, tmp2);
        return res;
    }
    void helper(vector<int>& curr_list, vector<int>& histort_index_list){
        if(curr_list.size()==MAX_LEN) res.push_back(curr_list);
        for(int i=0; i<MAX_LEN; i++){
            if(find(histort_index_list.begin(), histort_index_list.end(), i)!=histort_index_list.end()) continue;
            vector<int> tmp1 = curr_list;
            tmp1.push_back(nums_[i]);
            vector<int> tmp2(histort_index_list);
            tmp2.push_back(i);
            helper(tmp1, tmp2);
        }
    }
};